spanner/session.go

```diff
@@ -1107,6 +1208,79 @@ func (p *sessionPool) take(ctx context.Context) (*sessionHandle, error) {
 	}
 }
 
+// takeMultiplexed returns a cached session if there is available one; if there isn't
+// any, it tries to allocate a new one.
+func (p *sessionPool) takeMultiplexed(ctx context.Context) (*sessionHandle, error) {
+	if !p.enableMultiplexSession {
+		return p.take(ctx)
+	}
+	trace.TracePrintf(ctx, nil, "Acquiring a multiplexed session")
+	for {
+		var s *session
+
+		p.mu.Lock()
+		if !p.valid {
+			p.mu.Unlock()
+			return nil, errInvalidSessionPool
+		}
+		// use the multiplex session if it is available
+		if p.multiplexedSession != nil {
+			// Multiplexed session is available, get it.
+			s = p.multiplexedSession
+			trace.TracePrintf(ctx, map[string]interface{}{"sessionID": s.getID()},
+				"Acquired multiplexed session")
+		}
+		if s != nil {
+			p.mu.Unlock()
+			// From here, If healthcheck workers failed to
+			// schedule healthcheck for the session timely, do the check here.
+			// Because session check is still much cheaper than session
+			// creation, they should be reused as much as possible.
+			if !p.isHealthy(s) {
+				continue
+			}
```

Comment:
I don't think we should do this for a multiplexed session. It is considered healthy by definition as long as it is in the pool.